File: 0747-min-cost-climbing-stairs/solution.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        for (int i = 2; i < cost.size(); ++i)
            cost[i] = min(cost[i-1], cost[i-2]) + cost[i];
        return min(cost.back(), cost[cost.size() - 2]);
    }
};

File: 0747-min-cost-climbing-stairs/solution.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        memo = {}
        def dfs(index):
            if index <= 0:
                return 0
            if index in memo:
                return memo[index]
            memo[index] = min(dfs(index-1), dfs(index - 2)) + cost[index-1]
            return memo[index]

        return min(dfs(len(cost)), dfs(len(cost) - 1))
